### PR TITLE
Add quiver trace type for vector field visualization

### PR DESCRIPTION
**Pull request by degzhaus**
## Overview
This PR adds a new `quiver` trace type to Plotly.js for visualizing 2D vector fields using arrows.

Continuation of https://github.com/plotly/plotly.js/pull/7584

## Features
- **Vector positioning**: Arrows placed at `x`, `y` coordinates with direction/magnitude from `u`, `v` components
- **Sizing modes**: `scaled` (normalized), or `raw` vector lengths via 
  - Fixed scale factor can be applied via `sizeref`
- **Angle modes**:  Use `anglemode` to set arrow angle based on data axis scales (`data`) or visual angle on page
- **Arrow anchoring**: Position arrows by `tail`, `tip`, or `center`
- **Colorscale support**: Color arrows by magnitude or custom scalar field (`c` array)
- **Styling**: Configurable line color, width, dash patterns via `marker.line.*` attributes
- **Arrowhead sizing**: Adjustable arrowhead size via `marker.arrowsize`
- **Full interactivity**: Hover support with customizable `hovertemplate` and point selection
- **Animation support**: Animatable `x`, `y`, `u`, `v`, and `c` arrays

## API

```
Plotly.newPlot('div', [{
  type: 'quiver',
  x: [0, 1, 2],
  y: [0, 1, 2],
  u: [1, 0, -1],  // x-component of vectors
  v: [0, 1, 0],   // y-component of vectors
  sizemode: 'scaled',  //  'scaled' | 'raw'
  anglemode: 'data',  // 'data' | 'paper'
  sizeref: 0.5,
  anchor: 'tail',  // 'tail' | 'tip' | 'center'
  marker: {
    line: { width: 2 },
    colorscale: 'Viridis',
  },
}]);
```

## Screenshots

Examples taken from [plotly.com/python/quiver-plots](https://plotly.com/python/quiver-plots/)

[Gist with example code](https://gist.github.com/degzhaus/2fb3c08cbc228682125f94bce52ba7d1)

### Quiver Plot, vector field with mesh grid (colored by magnitude)
Screenshot 2026-07-20 at 6 41 08 PM

### Quiver plot, vector field with mesh grid (colored by independent color array)
Screenshot 2026-07-20 at 6 52 17 PM

## Testing
- ✅ Jasmine unit tests
- ✅ Image tests

### New files (`src/traces/quiver/`)
- `index.js` - Trace module definition
- `attributes.js` - Attribute schema
- `defaults.js` - Default value handling
- `calc.js` - Data calculation
- `plot.js` - SVG rendering
- `style.js` - Styling
- `hover.js` - Hover behavior
- `select_points.js` - Selection support
- `event_data.js` - Event data formatting
- `format_labels.js` - Label formatting

### Modified
- `lib/index.js` & `lib/index-strict.js` - Build integration

### Tests
- `test/jasmine/tests/quiver_test.js` - Unit tests
- `test/image/mocks/quiver_*.json` - New image mocks
- `test/image/baselines/quiver_*.png` - Baseline images

### Known issues
- Axes autorange when `anglemode` is `'paper'` is not quite correct, especially when vectors are long and/or axes aspect ratio is very different from plot aspect ratio.
- The squishing/stretching of existing arrows that happens during drag when resizing the plot area is correct when `anglemode` is `'data'`, but _incorrect_ when `anglemode` is `'paper'`. Arrows should not be squished/stretched in that case. There's no quiver-specific drag code that I can see, so I think the behavior is implemented at a lower level and we may need to override it under certain conditions. In any case, though, after the drag is released, the arrows adjust to their correct positions.
- The arrow-scaling heuristic based on point density works OK, but could probably be improved. In particular, for grids, as the size of the grid increases, arrow size approaches the distance between two adjacent grid points, so at the very least, the computed scale factor should probably be reduced by 5-10%. This heuristic probably also doesn't produce good results for very irregularly-spaced points, but I'm not sure what metric would be better in that case.
- Arrow scaling could be improved when a trace has just a single point, or all points share a single x/y value. Right now we just scale so that the longest vector has magnitude 1 in x*y space. For the case of a single data point, I'm not sure we can do much better, but for the case of multiple points in a line we can definitely choose a better scale factor (use "distance between points" rather than "point density").
- Log axes aren't really handled, but I'm not sure quiver traces on log axes make sense anyway. We should probably explicitly disallow log axes.
- Animations do not work (`Plotly.animate()` ultimately updates the plot with the new data, but the transition is not animated)

### Remaining TODOs (in addition to the above)
- Increase coverage of Jasmine and image tests
- Reflect on attribute naming for `anglemode` / `sizemode` / `sizeref`

### CODE REVIEW

**Comment by degzhaus** [comment]
Thanks again for the great comments, @emilykl ! 

I will attempt to reference commits inline from this comment here: 
https://github.com/plotly/plotly.js/pull/7584#issuecomment-3877802949

-- Bug Fixes --

"The axis ranges don't automatically adjust to include the full length of the arrows" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/96845a844e05d9ebc79106d8890450f95c88d0d8

"Colorscale seems broken in the latest version" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/3cd525c72ed01b5fa2fcf72e93f0954add519d45

"Legends are missing icons" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/6a3726026ed7cd2943651c5e3271709082f3ebf1

"`selected` and `unselected` styling options" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/96bb82cf87840cde2d262e34e3d382f2506f1d61

"The `text` property also seems broken in the latest iteration" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/69516dbc2536be6500ad9cd5bfc79d81c89e5136

"Display of colorbars does not seem to work" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/b19cce2d131618180c66966a9f3613085a449e38

-- API Refinements --

"`c` should become `marker.color`" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/3cd525c72ed01b5fa2fcf72e93f0954add519d45

"The line options should be nested under marker.line", 
"Move arrowheadsize under marker so that it becomes marker.arrowheadsize", 
"No need for duplicate arrowwidth property",
"arrow_scale should be removed as well" all fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/c7f073d9aacb72b129b89e8fdd15fb484debf17f

**Comment by degzhaus** [comment]
Here are the followups for the remaining 3 comments from: https://github.com/plotly/plotly.js/pull/7584, @emilykl :

"Could we plausibly reuse the src/traces/scatter/select.js selectPoints() function" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/4f91255b0e72d1ab7883b7ff1749b4c23924483b

"Should probably call the hasColorscale() function here" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/82e6bc117bb52d91796631eaeb92af629960fac2

"Can you look into whether we can/should use the handleXYDefaults() function here" fixed here:
https://github.com/plotly/plotly.js/pull/7710/changes/7135b5250f058fcaf93cf3ab250857f45698da3e

**Comment by degzhaus** [comment]
Alrighty, @emilykl , thanks again for your great feedback and patience as I addressed it!

I have pushed all my fixes and tested locally, I think we're close 🤞

**Comment by emilykl** [comment]
Great news @degzhaus ! Thank you for the continued work and the clear summary. I will try to take another look within the next week. Feel free to ping me next Tuesday if I haven't gotten back to you by then.

**Comment by degzhaus** [comment]
hey there, @emilykl, hope you are having a great start to the week!

i just rebased and pushed again 🥳

**Comment by camdecoster** [comment]
@degzhaus would you be able to merge/rebase one more time? Sorry for the extra work. We just changed how our test images are generated and that led to the baseline conflicts with this branch.

File: src/traces/quiver/attributes.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        editType: 'calc',
+        description: [
+            'Adjusts the arrow size scaling.',
+            'The arrow length is determined by the vector norm multiplied by `sizeref`,',
+            'optionally normalized when `sizemode` is *scaled*.'
+        ].join(' ')
+    },
+    anchor: {
+        valType: 'enumerated',
+        values: ['tip', 'tail', 'cm', 'center', 'middle'],
```
Comment (camdecoster) [inline src/traces/quiver/attributes.js]: Could you please remove `'cm'` and `'center'` as values? One term for the middle should be fine.

File: src/traces/quiver/attributes.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                description: 'Sets the text font color of unselected points, applied only when a selection exists.'
+            },
+            editType: 'style'
+        },
+        editType: 'style'
+    }
+};
+
+// Extend with base attributes (includes hoverinfo, etc.)
+extendFlat(attrs, baseAttrs);
```
Comment (camdecoster) [inline src/traces/quiver/attributes.js]: Did you mean to pasate `baseAttrs` on top of `attrs`? I think it would be safer to handle it this way. That's what is typically done when building attributes.

```suggestion
attrs = extendFlat({}, baseAttrs, attrs);
```

**Comment by emilykl** [comment]
Hi @degzhaus ! I've run down the [list of issues](https://github.com/plotly/plotly.js/pull/7584#issuecomment-3877802949) from my previous review, and most seem to be fixed 🥳 However I am still seeing these two issues:

> - The axis ranges don't automatically adjust to include the full length of the arrows. [See example here](https://codepen.io/emilykl-code/pen/vEKQvBg). By default if no axis range is specified in the layout, the entirety of all arrows should be visible.

The baseline images `quiver_anchor.png`, `quiver_arrow-styling.png`, `quiver_sizemode.png`, and `quiver_zero-vectors.png` show cut-off arrows.

> - Colorscale seems broken in the latest version, unless I'm missing something. The arrows in `quiver_colorscale.png` and `quiver_custom-colorscale.png` seem to be all the same color.

The baseline images  in `quiver_colorscale.png` and `quiver_custom-colorscale.png` are still all the same color. Also the `quiver_colorscale` should provide a value for `marker.color`; otherwise the arrows will not display a colorscale even if there was no bug.

I'm still doing a bit more testing so I will leave another comment if I find anything else. @camdecoster is taking a look as well, so keep an eye out for his comments.

Finally when you have a chance please merge/rebase against the master branch! We recently overhauled the CI pipeline and you will need to merge/rebase in order to run the new pipeline.

File: src/traces/quiver/style.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        d3.select(this).selectAll('path.js-line')
+            .call(Drawing.lineGroupStyle, markerLine.width, lineColor, markerLine.dash);
+    });
+}
+
+function styleOnSelect(gd, cd, sel) {
+    var trace = cd[0].trace;
+    var marker = trace.marker || {};
+    var markerLine = marker.line || {};
+    var lineColor = Lib.isArrayOrTypedArray(marker.color) ? undefined : marker.color;
```
Comment (camdecoster) [inline src/traces/quiver/style.js:46]: If `marker.color` is an array, `lineColor` will be undefined. This would result in the unselected color being undefined as well. Could you update the code to handle this situation and apply a dim version (via opacity) of the selected color?

File: src/traces/quiver/calc.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        Lib.mergeArray(trace.textfont.color, cd, 'tc');
+        Lib.mergeArray(trace.textfont.family, cd, 'tf');
+        Lib.mergeArray(trace.textfont.weight, cd, 'tw');
+        Lib.mergeArray(trace.textfont.style, cd, 'ty');
+        Lib.mergeArray(trace.textfont.variant, cd, 'tv');
+    }
+
+    // Colorscale cmin/cmax computation: prefer provided marker.color, else magnitude
+    if(trace._hasColorscale) {
+        var vals = hasMarkerColorArray ? [cMin, cMax] : [normMin, normMax];
```
Comment (camdecoster) [inline src/traces/quiver/calc.js:203]: Could you add a guard (here or elsewhere) that handles if `normMin` and `normMax` are infinite? This is an edge case that would require all invalid data, but we should guard against it.

File: test/jasmine/tests/quiver_test.js [hunk 5]
```diff
@@ -0,0 +1,319 @@
+var Plotly = require('../../../lib/index');
+var Lib = require('../../../src/lib');
+var d3Select = require('../../strict-d3').select;
```
Comment (camdecoster) [inline test/jasmine/tests/quiver_test.js]: Could you delete these unused imports?

**Comment by camdecoster** [comment]
@degzhaus are you still interested in finishing up this PR?

**Comment by degzhaus** [comment]
> @degzhaus are you still interested in finishing up this PR?

i am, and pardon my delay, Cam! i have begun working through the recent comments, hope to be ready for another round of review this week 🤞

**Comment by camdecoster** [comment]
Great to hear! Ping us when you're ready for another review.

**Comment by camdecoster** [comment]
Hello @degzhaus! We're working toward a new major release and we're interested in including this new trace as part of that. If we want that to happen, this PR will need to be merged in the next week or so. Do you think you'll have time to finish it up by then? We're happy to help if you need some assistance.

**Comment by degzhaus** [comment]
> test/jasmine/tests/quiver_test.js

Thank you so much for validating and calling these out, @emilykl !

Fix for Adjust axis ranges automatically to include full length of arrows:
https://github.com/plotly/plotly.js/pull/7710/changes/a2799e2a4a938c6498a0f4e7f7a8661602d9a91f

Fix for Colorscale:
https://github.com/plotly/plotly.js/pull/7710/changes/334486499244ce165b2413902a3dec3dda849e00

File: test/jasmine/tests/quiver_test.js (hunk 5 above)
Comment (degzhaus) [reply test/jasmine/tests/quiver_test.js]: Absolutely, removed here:

https://github.com/plotly/plotly.js/pull/7710/changes/eba15eb1fc19dde1ac54802a1c9f95219a710dc4

File: src/traces/quiver/calc.js (hunk 4 above)
Comment (degzhaus) [reply src/traces/quiver/calc.js:203]: Good callout, thank you, Cam, added guard here:

https://github.com/plotly/plotly.js/pull/7710/changes/dd71b9435c2a687e27d09126b9a3939914784231

File: src/traces/quiver/style.js (hunk 3 above)
Comment (degzhaus) [reply src/traces/quiver/style.js:46]: Thanks for pointing this out. I was not able to figure out exactly how to do that with opacity, so I styled the path with `stroke-opacity` here:

https://github.com/plotly/plotly.js/pull/7710/changes/f3789e6332aad4d126fc3d1b25ea96c1c5ce3f2d

Happy to adjust if you have something else in mind!

File: src/traces/quiver/attributes.js (hunk 2 above)
Comment (degzhaus) [reply src/traces/quiver/attributes.js]: Good call, thank you, changed here:

https://github.com/plotly/plotly.js/pull/7710/changes/2e214c7deb7790e7064ae31230f9235fcb0f540e

File: src/traces/quiver/attributes.js (hunk 1 above)
Comment (degzhaus) [reply src/traces/quiver/attributes.js]: Much simpler, thanks, removed redundant anchor value options:

https://github.com/plotly/plotly.js/pull/7710/changes/94dc00d952c894c6d0761337c78f64a33cc10fe2

**Comment by degzhaus** [comment]
> Hello @degzhaus! We're working toward a new major release and we're interested in including this new trace as part of that. If we want that to happen, this PR will need to be merged in the next week or so. Do you think you'll have time to finish it up by then? We're happy to help if you need some assistance.

Amazing, that is so exciting! Thanks so much for y'all's direction, @emilykl and @camdecoster, I learned a lot working with each of you.

I responded inline with what I think are the remaining outstanding points. Very excited for final steps.

Thanks so much for reaching out to be sure this is included in the release, that means a lot.

Feel free to ping with any remaining tasks, I will keep my eyes peeled for updates.

All the best.

**Comment by camdecoster** [comment]
I believe that you've addressed all of my comments. I'll let @emilykl take a look and give the final review. I'm going to edit this PR to target our 4.0 branch instead of master. Then I'll merge the 4.0 changes into your branch and cross my fingers that it just works. I'll push those changes and then it should good to go for Emily to look at.

**Comment by camdecoster** [comment]
Hey! It worked! It looks like you just need to update the baseline images, address the failing test, and add a draftlog entry (plus whatever Emily comments on). Almost there!

**Comment by degzhaus** [comment]
> Hey! It worked! It looks like you just need to update the baseline images, address the failing test, and add a draftlog entry (plus whatever Emily comments on). Almost there!

Great news, and thank you so much for hopping on this today and taking care of those requisite commits, @camdecoster !

I just pushed 3 more commits as well, hopefully CI will be as happy as me 🥳

Is there something I can do to initiate the pipeline or will that happen automatically?

Thanks again for everything, Cam and Emily 🙌

**Comment by camdecoster** [comment]
I approved the workflows. All green.

**Comment by emilykl** [comment]
@degzhaus Exciting news! I'm starting to review, I'll get back to you ASAP with any comments.

File: test/image/mocks/quiver_simple.json
Comment (emilykl) [inline test/image/mocks/quiver_simple.json:1]: @degzhaus I think it would be OK to remove this image test (and the associated PNG) since as far as I can tell it doesn't test any properties which aren't covered by the other mocks.

File: test/image/mocks/quiver_zero-vectors.json
Comment (emilykl) [inline test/image/mocks/quiver_zero-vectors.json:1]: @degzhaus Instead of testing this in a separate mock, could you edit one of the other mocks to include zero-length vectors and remove this one? 

(Each image test adds a bit of CI time, so we want to strike a balance between having comprehensive tests while also keeping the number of image tests down.)

File: test/jasmine/assets/mock_lists.js [hunk 6]
```diff
@@ -28,6 +28,7 @@ var svgMockList = [
     ['layout-colorway', require('../../image/mocks/layout-colorway.json')],
     ['multicategory', require('../../image/mocks/multicategory.json')],
     ['polar_categories', require('../../image/mocks/polar_categories.json')],
+    ['quiver_simple', require('../../image/mocks/quiver_simple.json')],
```
Comment (emilykl) [inline test/jasmine/assets/mock_lists.js]: I don't think this is needed

```suggestion
```

File: test/jasmine/tests/lib_test.js [hunk 7]
```diff
@@ -2784,8 +2784,9 @@ describe('Test lib.js:', function () {
             for (var i = 0; i < 15; i++) {
                 Lib.hovertemplateString({ fallback: '', template: '%{idontexist}' });
             }
-            // Expect 11 since the suppression warning also calls Lib.warn
-            expect(Lib.warn.calls.count()).toBe(11);
+            // Depending on prior tests, the suppression warning may already be reached once.
+            expect(Lib.warn.calls.count()).toBeGreaterThanOrEqual(10);
+            expect(Lib.warn.calls.count()).toBeLessThanOrEqual(11);
```
Comment (emilykl) [inline test/jasmine/tests/lib_test.js:2789]: @camdecoster what do you think about this patch for this flaky test? It loosens the test, but we already re-run until it passes anyway, so maybe it's fine to just loosen the test a bit to maintain our sanity.

File: src/traces/quiver/attributes.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            'or *center* to center the arrow on (x,y).'
+        ].join(' ')
+    },
+    hoverdistance: {
+        valType: 'number',
+        min: -1,
+        dflt: 20,
+        editType: 'calc',
+        description: 'Maximum distance (in pixels) to look for nearby arrows on hover.'
+    },
```
Comment (emilykl) [inline src/traces/quiver/attributes.js]: No other trace type has a `hoverdistance` attribute (it's a layout property) so I'd suggest to remove it unless there's a good reason to define it at the trace level here.

```suggestion
```

File: test/jasmine/tests/lib_test.js (hunk 7 above)
Comment (camdecoster) [reply test/jasmine/tests/lib_test.js:2789]: I'm okay with it. I ran into that test with the `hovertemplate` work I did. It's not a great test and should be updated when we switch to a new testing library.

File: src/traces/quiver/attributes.js [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        editType: 'calc',
+        description: 'Sets the positions of the `text` elements with respects to the (x,y) coordinates.'
+    },
+    // Text font
+    textfont: fontAttrs({
+        editType: 'calc',
+        colorEditType: 'style',
+        arrayOk: true,
+        description: 'Sets the text font.'
+    }),
```
Comment (emilykl) [inline src/traces/quiver/attributes.js]: Preference for reusing the scatter attributes here if possible. Can use `extendFlat` if a few attribute properties need to be changed.

Of course if the whole attribute definition really does need to be different, it's OK to redefine here.

```suggestion
    text: scatterAttrs.text,
    textposition: scatterAttrs.textposition,
    textfont: scatterAttrs.textfont,
```

File: src/traces/quiver/attributes.js [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+// We need to create a new object to avoid mutating the shared base attributes
+attrs.hoverinfo = extendFlat({}, baseAttrs.hoverinfo, {
+    flags: ['x', 'y', 'u', 'v', 'text', 'name'],
+    dflt: 'all'
+});
+
+// Add hovertemplate
+attrs.hovertemplate = extendFlat({}, hovertemplateAttrs({}, {
+    keys: ['x', 'y', 'u', 'v', 'text', 'name']
+}));
```
Comment (emilykl) [inline src/traces/quiver/attributes.js]: @degzhaus Is there a reason why these are added afterwards, rather than defined in the main body of `attrs`?

File: src/traces/scatter/select.js [hunk 11]
```diff
@@ -13,7 +13,7 @@ module.exports = function selectPoints(searchInfo, selectionTester) {
     var x;
     var y;
 
-    var hasOnlyLines = (!subtypes.hasMarkers(trace) && !subtypes.hasText(trace));
+    var hasOnlyLines = trace.mode && !subtypes.hasMarkers(trace) && !subtypes.hasText(trace);
```
Comment (emilykl) [inline src/traces/scatter/select.js:16]: IMO it would be clearer to modify the `subtypes.hasMarkers(trace)` logic such that it returns `true` for quiver traces, rather than adding a `trace.mode` escape hatch here.

**Comment by emilykl** [comment]
FYI @degzhaus I've left some comments but haven't quite finished my review yet, I'll continue on Monday!

File: src/traces/quiver/calc.js [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+            if(norm < normMin) normMin = norm;
+        }
+
+        if(hasMarkerColorArray) {
+            var ci = markerColor[i];
+            if(isNumeric(ci)) {
+                if(ci < cMin) cMin = ci;
+                if(ci > cMax) cMax = ci;
+            }
+        }
```
Comment (emilykl) [inline src/traces/quiver/calc.js]: I think _all_ of this logic (lines 49-64) should be inside the `if(xValid && yValid)` case, no? The vector norm should not count towards `_maxNorm` unless its x and y values are valid (meaning it can be plotted).

File: src/traces/quiver/calc.js [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        if(xValid && yValid) {
+            cdi.x = xVals[i];
+            cdi.y = yVals[i];
+        } else {
+            cdi.x = BADNUM;
+            cdi.y = BADNUM;
+        }
+
+        var ui = uArr[i] || 0;
+        var vi = vArr[i] || 0;
```
Comment (emilykl) [inline src/traces/quiver/calc.js]: We should probably check that `u` and `v` are valid as well, right?

File: src/traces/scatter/select.js (hunk 11 above)
Comment (emilykl) [reply src/traces/scatter/select.js:16]: @degzhaus actually, scratch my previous comment, I've realized that modifying `hasMarkers()` has a larger blast radius since it's used in ways that wouldn't apply to quiver traces.

This logic can probably be improved but it's OK to leave as-is for now.

**Comment by emilykl** [comment]
@degzhaus I've pushed 3 commits to the branch [`quiver-edits-ekl`](https://github.com/plotly/plotly.js/tree/quiver-edits-ekl) to ensure the selection buttons are shown in the modebar, and fix the behavior of the `selected` and `unselected` styling attributes. Can you pull those commits onto your branch?

There's an additional bug where the arrowheads are much too large on zooming in — looks like they keep their same dimensions in _data_ coordinates, rather than keeping the same pixel size. If you're OK with it, I'll push a fix tomorrow which you can pull onto your branch.

**Comment by degzhaus** [comment]
> @degzhaus I've pushed 3 commits to the branch [`quiver-edits-ekl`](https://github.com/plotly/plotly.js/tree/quiver-edits-ekl) to ensure the selection buttons are shown in the modebar, and fix the behavior of the `selected` and `unselected` styling attributes. Can you pull those commits onto your branch?
> 
> There's an additional bug where the arrowheads are much too large on zooming in — looks like they keep their same dimensions in _data_ coordinates, rather than keeping the same pixel size. If you're OK with it, I'll push a fix tomorrow which you can pull onto your branch.

Thank you so much, @emilykl ! I started working through your other comments, will make some more time this afternoon to hopefully get through the rest. Sounds good on the arrowhead zooming bug. Feel free to push directly to my branch, I can pull before pushing.

Appreciate your help in getting this across the finish line!

**Comment by emilykl** [comment]
Sounds good @degzhaus !

I just pushed the 3 previously-mentioned commits directly to your branch, plus one more which fixes the arrowhead sizing. You can take a look and make sure it looks good to you. I'll let you work through the remaining comments but I'm happy to make some of those changes if you like. I'm continuing to do some more manual testing today so I'll let you know if I find any other issues.

Thank _you_ for your work on this trace!

File: src/traces/quiver/calc.js (hunk 13 above)
Comment (degzhaus) [reply src/traces/quiver/calc.js]: good call, thank you, added here:

https://github.com/plotly/plotly.js/pull/7710/changes/f49782051d8b0275fe936104e2f6bd066c22e00a

File: src/traces/quiver/calc.js (hunk 12 above)
Comment (degzhaus) [reply src/traces/quiver/calc.js]: nice catch, thank you, moved here:

https://github.com/plotly/plotly.js/pull/7710/changes/0f5069d57f3549217c5ef5fff1ec89f82351afa8

File: src/traces/quiver/attributes.js (hunk 10 above)
Comment (degzhaus) [reply src/traces/quiver/attributes.js]: good idea, thanks, Emily, changed here:

https://github.com/plotly/plotly.js/pull/7710/changes/2fc0c66c07f5867974cd93514470c64b6e786ecb

File: src/traces/quiver/attributes.js (hunk 9 above)
Comment (degzhaus) [reply src/traces/quiver/attributes.js]: much cleaner, thank you, changed here:

https://github.com/plotly/plotly.js/pull/7710/changes/8b1f4069410bfc93a4e52ac92ab2f0a11eeac764

File: src/traces/quiver/attributes.js (hunk 8 above)
Comment (degzhaus) [reply src/traces/quiver/attributes.js]: good catch, removed those here:

https://github.com/plotly/plotly.js/pull/7710/changes/232eda6c65603f29f7fa1871d9e4134d9282748a

File: test/image/mocks/quiver_zero-vectors.json
Comment (degzhaus) [reply test/image/mocks/quiver_zero-vectors.json:1]: good idea, thanks Emily, folded moved those here:

https://github.com/plotly/plotly.js/pull/7710/changes/aad037a189459da4b2879aae130038e5348d0b2d

File: test/image/mocks/quiver_simple.json
Comment (degzhaus) [reply test/image/mocks/quiver_simple.json:1]: nice, i like it, thanks Emily, rearranged here:

https://github.com/plotly/plotly.js/pull/7710/changes/ad4e647d545488d1a20110477e782cdbf2b0cc18

**Comment by degzhaus** [comment]
@emilykl Thanks again for your diligence and help getting this across the finish line, especially the commits you've been pushing!

I believe I have addressed all the open comments that you have not already addressed.

Looking forward to next steps!

(cc @camdecoster)

**Comment by emilykl** [comment]
Thanks @degzhaus, the updates look good! Doing a final review pass now, I might push a few additional commits if needed.

File: draftlogs/7710_add.md [hunk 14]
```diff
@@ -0,0 +1 @@
+- Add `quiver` trace type to visualize vector fields using arrows [[#7710](https://github.com/plotly/plotly.js/pull/7710)]
```
Comment (camdecoster) [inline draftlogs/7710_add.md]:
```suggestion
- Add `quiver` trace type to visualize vector fields using arrows [[#7710](https://github.com/plotly/plotly.js/pull/7710)], with thanks to @degzhaus for the contribution!
```

Comment (emilykl) [inline draftlogs/7710_add.md]:
```suggestion
- Add `quiver` trace type to visualize vector fields using arrows [[#7710](https://github.com/plotly/plotly.js/pull/7710)], with thanks to @degzhaus for the contribution!
```

**Review by camdecoster — APPROVED** [review]
Looks good to me!

**Comment by gpdf** [comment]
I’m excited - are we close to merging, and is it looking like we’re going to make it into v4?

**Comment by emilykl** [comment]
> I’m excited - are we close to merging, and is it looking like we’re going to make it into v4?

@degzhaus Yes! There are a still a few outstanding issues I'm aware of, which I've noted in the PR description, but we're planning to release a v4.0 release candidate imminently, so we'd like to go ahead and merge this PR and address remaining issues in a follow-up PR. That means we're on track for quiver to make it into 4.0.

**Comment by emilykl** [comment]
@degzhaus Congratulations, we've merged this PR into the `v4.0` branch! 🥳 🎉 

We'll be putting out a release candidate for v4.0 very shortly.

That doesn't mean no more changes can be made: the quiver API can be modified up until the v4.0 full release, and bug fixes can happen anytime. If you find anything that should be changed, go ahead and open up an issue, and a PR into `v4.0` if you like. We'll continue working through any issues we identify as well.

Thanks again for all your work on this trace type, it's been a lot of fun collaborating and we're excited to get quiver out to the public in the near future.

**Comment by degzhaus** [comment]
> @degzhaus Congratulations, we've merged this PR into the `v4.0` branch! 🥳 🎉
> 
> We'll be putting out a release candidate for v4.0 very shortly.
> 
> That doesn't mean no more changes can be made: the quiver API can be modified up until the v4.0 full release, and bug fixes can happen anytime. If you find anything that should be changed, go ahead and open up an issue, and a PR into `v4.0` if you like. We'll continue working through any issues we identify as well.
> 
> Thanks again for all your work on this trace type, it's been a lot of fun collaborating and we're excited to get quiver out to the public in the near future.

Eeeeek, so exciting! Thank you so much, @emilykl for all your mentorship along the way, I learned so much on this project. And thank you @camdecoster for your help and direction as well. Really great working with you both!

And thank you so much for your patience and enthusiasm @gpdf! Hope this is useful to you and your team.

All the best,

John

**Comment by AnnMarieW** [comment]
@degzhaus @emilykl @camdecoster @gpdf 

Congrats on getting this merged - This is so exciting!

There is a Plotly community forum post about this feature if you would like to chime in over there :slightly_smiling_face: 
https://community.plotly.com/t/quiver-plots-preview-new/97450/
